File: app/src/main/java/com/aminography/taskmanagerapp/TaskLifecycleOwner.java
package com.aminography.taskmanagerapp;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.os.Bundle;
import android.os.Handler;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LifecycleRegistry;

/**
 * @author aminography
 */
class TaskLifecycleOwner implements Application.ActivityLifecycleCallbacks, LifecycleOwner {

    private final Class<?> activityClass;
    private int targetTaskId = -1;

    private static final long TIMEOUT_MS = 700;

    // ground truth counters
    private int mStartedCounter = 0;
    private int mResumedCounter = 0;

    private boolean mPauseSent = true;
    private boolean mStopSent = true;

    private final Handler mHandler = new Handler();
    private final LifecycleRegistry mRegistry = new LifecycleRegistry(this);

    // To distinguish between screen rotation and normal lifecycle changes:
    private final Runnable mDelayedPauseRunnable = new Runnable() {
        @Override
        public void run() {
            dispatchPauseIfNeeded();
            dispatchStopIfNeeded();
        }
    };

    private static TaskLifecycleOwner sInstance = null;

    public static TaskLifecycleOwner getInstance(Context context, Class<?> activityClass) {
        if (sInstance == null) {
            sInstance = new TaskLifecycleOwner(activityClass);
            sInstance.attach(context);
        }
        return sInstance;
    }

    public static void destroyInstance(Context context) {
        if (sInstance != null) {
            sInstance.detach(context);
            sInstance = null;
        }
    }

    private TaskLifecycleOwner(Class<?> activityClass) {
        this.activityClass = activityClass;
    }

    @Override
    public void onActivityCreated(@NonNull Activity activity, @Nullable Bundle bundle) {
        if (activityClass.isInstance(activity)) {
            targetTaskId = activity.getTaskId();
        }
    }

    @Override
    public void onActivityStarted(@NonNull Activity activity) {
        if (activity.getTaskId() == targetTaskId) {
            activityStarted();
        }
    }

    @Override
    public void onActivityResumed(@NonNull Activity activity) {
        if (activity.getTaskId() == targetTaskId) {
            activityResumed();
        }
    }

    @Override
    public void onActivityPaused(@NonNull Activity activity) {
        if (activity.getTaskId() == targetTaskId) {
            activityPaused();
        }
    }

    @Override
    public void onActivityStopped(@NonNull Activity activity) {
        if (activity.getTaskId() == targetTaskId) {
            activityStopped();
        }
    }

    @Override
    public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle bundle) {
    }

    @Override
    public void onActivityDestroyed(@NonNull Activity activity) {
    }

    @NonNull
    @Override
    public Lifecycle getLifecycle() {
        return mRegistry;
    }

    private void activityStarted() {
        mStartedCounter++;
        if (mStartedCounter == 1 && mStopSent) {
            mRegistry.handleLifecycleEvent(Lifecycle.Event.ON_START);
            mStopSent = false;
        }
    }

    private void activityResumed() {
        mResumedCounter++;
        if (mResumedCounter == 1) {
            if (mPauseSent) {
                mRegistry.handleLifecycleEvent(Lifecycle.Event.ON_RESUME);
                mPauseSent = false;
            } else {
                mHandler.removeCallbacks(mDelayedPauseRunnable);
            }
        }
    }

    private void activityPaused() {
        mResumedCounter--;
        if (mResumedCounter == 0) {
            mHandler.postDelayed(mDelayedPauseRunnable, TIMEOUT_MS);
        }
    }

    private void activityStopped() {
        mStartedCounter--;
        dispatchStopIfNeeded();
    }

    private void dispatchPauseIfNeeded() {
        if (mResumedCounter == 0) {
            mPauseSent = true;
            mRegistry.handleLifecycleEvent(Lifecycle.Event.ON_PAUSE);
        }
    }

    private void dispatchStopIfNeeded() {
        if (mStartedCounter == 0 && mPauseSent) {
            mRegistry.handleLifecycleEvent(Lifecycle.Event.ON_STOP);
            mStopSent = true;
        }
    }

    private void attach(Context context) {
        mRegistry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE);
        ((Application) context.getApplicationContext()).registerActivityLifecycleCallbacks(this);
    }

    private void detach(Context context) {
        ((Application) context.getApplicationContext()).unregisterActivityLifecycleCallbacks(this);
        mRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY);
    }
}

File: app/src/main/java/com/aminography/taskmanagerapp/TaskLifecycleOwner.kt
package com.aminography.taskmanagerapp

import android.app.Activity
import android.app.Application
import android.content.Context
import android.os.Bundle
import android.os.Handler
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LifecycleRegistry

/**
 * @author aminography
 */
class TaskLifecycleOwner private constructor(
    private val activityClass: Class<*>
) : Application.ActivityLifecycleCallbacks, LifecycleOwner {

    private var targetTaskId = -1

    private var startedCounter = 0
    private var resumedCounter = 0

    private var pauseSent = true
    private var stopSent = true

    private val handler = Handler()

    private val delayedPauseRunnable = Runnable {
        dispatchPauseIfNeeded()
        dispatchStopIfNeeded()
    }

    private val registry = LifecycleRegistry(this)

    override fun getLifecycle(): Lifecycle = registry

    override fun onActivityCreated(activity: Activity, bundle: Bundle?) {
        if (activityClass.isInstance(activity)) {
            targetTaskId = activity.taskId
        }
    }

    override fun onActivityStarted(activity: Activity) {
        if (activity.taskId == targetTaskId) {
            activityStarted()
        }
    }

    override fun onActivityResumed(activity: Activity) {
        if (activity.taskId == targetTaskId) {
            activityResumed()
        }
    }

    override fun onActivityPaused(activity: Activity) {
        if (activity.taskId == targetTaskId) {
            activityPaused()
        }
    }

    override fun onActivityStopped(activity: Activity) {
        if (activity.taskId == targetTaskId) {
            activityStopped()
        }
    }

    override fun onActivitySaveInstanceState(activity: Activity, bundle: Bundle) {
    }

    override fun onActivityDestroyed(activity: Activity) {
    }

    private fun activityStarted() {
        startedCounter++
        if (startedCounter == 1 && stopSent) {
            registry.handleLifecycleEvent(Lifecycle.Event.ON_START)
            stopSent = false
        }
    }

    private fun activityResumed() {
        resumedCounter++
        if (resumedCounter == 1) {
            if (pauseSent) {
                registry.handleLifecycleEvent(Lifecycle.Event.ON_RESUME)
                pauseSent = false
            } else {
                handler.removeCallbacks(delayedPauseRunnable)
            }
        }
    }

    private fun activityPaused() {
        resumedCounter--
        if (resumedCounter == 0) {
            handler.postDelayed(delayedPauseRunnable, TIMEOUT_MS)
        }
    }

    private fun activityStopped() {
        startedCounter--
        dispatchStopIfNeeded()
    }

    private fun dispatchPauseIfNeeded() {
        if (resumedCounter == 0) {
            pauseSent = true
            registry.handleLifecycleEvent(Lifecycle.Event.ON_PAUSE)
        }
    }

    private fun dispatchStopIfNeeded() {
        if (startedCounter == 0 && pauseSent) {
            registry.handleLifecycleEvent(Lifecycle.Event.ON_STOP)
            stopSent = true
        }
    }

    private fun attach(context: Context) {
        registry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE)
        (context.applicationContext as Application).registerActivityLifecycleCallbacks(this)
    }

    private fun detach(context: Context) {
        (context.applicationContext as Application).unregisterActivityLifecycleCallbacks(this)
        registry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    }

    companion object {
        private const val TIMEOUT_MS = 700L

        private var instance: TaskLifecycleOwner? = null

        fun getInstance(context: Context, activityClass: Class<*>): TaskLifecycleOwner {
            if (instance == null) {
                instance = TaskLifecycleOwner(activityClass).apply { attach(context) }
            }
            return requireNotNull(instance)
        }

        fun destroy(context: Context) {
            instance?.detach(context)
            instance = null
        }
    }
}
